src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs

```diff
@@ -88,5 +126,68 @@ internal bool IsTagHelper(INamedTypeSymbol symbol)
                     symbol.AllInterfaces.Contains(_interface);
             }
         }
+
+        internal class ViewComponentVisitor : SymbolVisitor
+        {
+            private INamedTypeSymbol _interface;
+            private List<INamedTypeSymbol> _results;
+
+            public ViewComponentVisitor(INamedTypeSymbol @interface, List<INamedTypeSymbol> results)
+            {
+                _interface = @interface;
+                _results = results;
+            }
+
+            public override void VisitNamedType(INamedTypeSymbol symbol)
+            {
+                if (IsViewComponent(symbol))
+                {
+                    _results.Add(symbol);
+                }
+
+                foreach (var member in symbol.GetTypeMembers())
+                {
+                    Visit(member);
+                }
+            }
+
+            public override void VisitNamespace(INamespaceSymbol symbol)
+            {
+                foreach (var member in symbol.GetMembers())
+                {
+                    Visit(member);
+                }
+            }
+
+            internal bool IsViewComponent(INamedTypeSymbol symbol)
+            {
+                if (symbol.DeclaredAccessibility != Accessibility.Public ||
+                    symbol.IsAbstract ||
+                    symbol.IsGenericType)
+                {
+                    return false;
+                }
+
+                return symbol.Name.EndsWith(ViewComponentTypes.ViewComponentSuffix) ||
```

Comment:
There's also `NonViewComponent` attribute to consider, which makes the entire hierarchy not a VC.